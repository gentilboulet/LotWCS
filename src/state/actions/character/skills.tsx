import { createAction } from "typesafe-actions";

import { TSkillName } from "../../../data/skills";
import { ICost } from "../../models/character/costs";
import { historyMetaCreator } from "../meta";

export const skillsBuy = createAction(
  "skills/BUY",
  (name: TSkillName, cost: ICost) => ({ cost, name }),
  historyMetaCreator,
)();

export const skillSpecialityBuy = createAction(
  "skills/BUY_SPECIALITY",
  (skill: TSkillName, speciality: string, cost: ICost) => {
    if (speciality.length === 0) {
      throw new Error("Empty speciality");
    }
    return { cost, skill, speciality };
  },
  historyMetaCreator,
)();
