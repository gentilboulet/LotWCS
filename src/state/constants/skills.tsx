export const SKILLS_BUY = 'SKILLS_BUY';
export type SKILLS_BUY = typeof SKILLS_BUY;

export const SKILLS_SPECIALITY_BUY = 'SKILLS_SPECIALITY_BUY';
export type SKILLS_SPECIALITY_BUY = typeof SKILLS_SPECIALITY_BUY;
