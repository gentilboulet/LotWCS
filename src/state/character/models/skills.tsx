import { skills as data, TSkillName } from "../../../data/skills";
import { maxSkillBonus } from "./derived";
import { ICharacterState } from "./type";

interface ISkill {
  name: TSkillName;
  specialities: string[];
  value: number;
}

export type TSkillsState = ISkill[];

export function createState(): TSkillsState {
  return Object.keys(data).map((name: TSkillName) => {
    return { name, specialities: [], value: 0 };
  });
}

export function increase(state: TSkillsState, skillName: TSkillName): void {
  getSkill(state, skillName).value += 5;
}

export function isSpecialityPresent(
  state: TSkillsState,
  skillName: TSkillName,
  specialityName: string,
): boolean {
  const specialityIndex = (state.find(
    s => s.name === skillName,
  ) as ISkill).specialities.findIndex(
    (speciality: string) => speciality === specialityName,
  );
  return specialityIndex !== -1;
}

export function addSpeciality(
  state: TSkillsState,
  skillName: TSkillName,
  specialityName: string,
): void {
  if (isSpecialityPresent(state, skillName, specialityName)) {
    throw new Error(
      'Something went wrong, speciality "' +
        specialityName +
        '" for skill "' +
        skillName +
        '" already bought',
    );
  }
  // getSkill(state, skillName).specialities.push(specialityName);
  const spe = getSkill(state, skillName);
  spe.specialities.push(specialityName);
  if (typeof specialityName != "string")
    throw new Error("Only stringm not" + specialityName);
}

export function canBuySkill(
  state: ICharacterState,
  skillName: TSkillName,
): boolean {
  const value = getSkill(state.skills, skillName).value;
  const max = maxSkillBonus(state);
  return !(value + 5 > max);
}

export function canBuySpeciality(
  state: ICharacterState,
  skillName: TSkillName,
  speciality: string,
): boolean {
  return !isSpecialityPresent(state.skills, skillName, speciality);
}

export function getSkill(state: TSkillsState, skillName: TSkillName) {
  console.log("searching for", skillName);
  return state.find(s => s.name === skillName) as ISkill;
}
