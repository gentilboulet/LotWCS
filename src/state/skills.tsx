import { skills as data, TSkillName } from 'data/skills';

import { canPayCost, getCostSkill, getCostSpeciality } from 'state/costs';
import { maxSkillBonus } from 'state/derived';
import { IStoreState } from 'state/type';

interface ISkill {
  name: string;
  specialities: string[];
  value: number;
}

export type TSkillsState = {
  [skill in TSkillName]: ISkill;
}

export function createState(): TSkillsState {
  const state = {};
  Object.keys(data).forEach(skill => {
    state[skill] = {
      name: skill,
      specialities: [],
      value: 0
    };
  });
  return state as TSkillsState;
}

export function increase(state: TSkillsState, skillName: string, maxSkillValue: number): void {
  const old = state[skillName].value;
  if( old + 5 > maxSkillValue) { throw new Error('Skill '+skillName+' overflow'); }
  state[skillName].value += 5;
}

export function isSpecialityPresent(state: TSkillsState, skillName: string, specialityName: string): boolean {
  const specialityIndex = state[skillName].specialities.findIndex((speciality: string) => speciality === specialityName);
  return specialityIndex !== -1;
}

export function addSpeciality(state: TSkillsState, skillName: string, specialityName: string): void {
  if(isSpecialityPresent(state, skillName, specialityName) ) {
    throw new Error('Something went wrong, speciality "'+specialityName+'" for skill "'+skillName+'" already bought');
  }
  state[skillName].specialities.push(specialityName);
}

export function canBuySkill(state: IStoreState, skillName: TSkillName): boolean {
  const value = state.skills[skillName].value;
  const max = maxSkillBonus(state);
  if ( (value + 5) > max ) { return false; }
  const cost = getCostSkill(state, skillName);
  return canPayCost(state, cost);
}

export function canBuySpeciality(state: IStoreState, skillName: TSkillName, speciality: string): boolean {
  if ( isSpecialityPresent(state.skills, skillName, speciality) ) { return false; }
  const cost = getCostSpeciality(state, skillName, speciality);
  return canPayCost(state, cost);
}
