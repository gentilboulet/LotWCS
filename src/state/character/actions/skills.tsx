import { createAction } from "typesafe-actions";

import { TSkillName } from "../../../data/skills";
import { ICost } from "../models/costs";

export const skillsBuy = createAction(
  "skills/BUY",
  (name: TSkillName, cost: ICost) => ({ cost, name }),
)();

export const skillSpecialityBuy = createAction(
  "skills/BUY_SPECIALITY",
  (skill: TSkillName, speciality: string, cost: ICost) => {
    if (speciality.length === 0) {
      throw new Error("Empty speciality");
    }
    return { cost, skill, speciality };
  },
)();
