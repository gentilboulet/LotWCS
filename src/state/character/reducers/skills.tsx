import produce from "immer";
import { ActionType, getType } from "typesafe-actions";

import * as actions from "../actions/skills";
import { addSpeciality, increase, TSkillsState } from "../models/skills";

export function skillsReducer(
  state: TSkillsState,
  action: ActionType<typeof actions>,
): TSkillsState {
  return produce(state, draft => {
    switch (action.type) {
      case getType(actions.skillsBuy):
        increase(draft, action.payload.name);
        break;
      case getType(actions.skillSpecialityBuy):
        addSpeciality(draft, action.payload.skill, action.payload.speciality);
        break;
      default:
    }
  });
}
