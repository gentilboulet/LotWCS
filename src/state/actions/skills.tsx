import { TSkillName } from 'data/skills';
import * as constants from 'state/constants/skills';
import { ICost } from 'state/costs';


export interface ISkillBuy {
  type: constants.SKILLS_BUY;
  name: string;
  cost: ICost;
}

export interface ISkillSpecialityBuy {
  type: constants.SKILLS_SPECIALITY_BUY;
  skill: string;
  speciality: string;
  cost: ICost;
}

export type ISkillAction = ISkillBuy | ISkillSpecialityBuy;

export function skillsBuy(name: TSkillName, cost: ICost): ISkillBuy {
  return { cost, name, type: constants.SKILLS_BUY };
}

export function skillSpecialityBuy(skill: TSkillName, speciality: string, cost: ICost): ISkillSpecialityBuy {
  if (speciality.length === 0) { throw new Error ('Empty speciality'); }
  return { cost, skill, speciality, type: constants.SKILLS_SPECIALITY_BUY };
}
