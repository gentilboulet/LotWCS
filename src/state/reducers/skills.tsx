import { ISkillAction } from 'state/actions/skills';
import * as constants from 'constants/skills';
import * as derived from 'state/derived';
import { IStoreSkillJS, IStoreSkillSpecialityJS, IStoreState, specialityFactory } from 'state/types';

// Sub Reducers
import { applyCost } from 'costs/reducer';
import { pushToHistory } from './history';

export function getSkillIndex(state: IStoreState, skillName: string): number {
 return state.get('skills').findIndex((s: IStoreSkillJS) => (s.name === skillName));
}

export function increaseValue(state: IStoreState, skillName: string): void {
  const index = getSkillIndex(state, skillName);

  if (state.getIn(['skills', index]).value + 5 >  derived.maxSkillBonus(state)) {
    throw new Error('Something went wrong, skill overflow');
  }

  state.updateIn(['skills', index, 'value'], v => (v + 5));
}

export function getSpecialityIndex(state: IStoreState, skillName: string, specialityName: string): number {
 const skillIndex = getSkillIndex(state, skillName);

 if (skillIndex === -1) { throw new Error('Something went wrong, unknown skill for speciality'); }

 return state.get('skillSpecialities').findIndex((speciality: IStoreSkillSpecialityJS) => {
   return speciality.name === specialityName;
  });
}

export function addSpeciality(state: IStoreState, skillName: string,  specialityName: string): void {
 const specialityIndex = getSpecialityIndex(state, skillName, specialityName);

 if (specialityIndex !== -1) {
  throw new Error('Something went wrong, speciality already bought');
 }

 state.updateIn(['skillSpecialities'], list => list.push(
     specialityFactory({ skill: skillName, name: specialityName})
   )
 );
}

export function skillsReducer(oldState: IStoreState, action: ISkillAction): IStoreState {
  switch (action.type) {
    case constants.SKILLS_BUY:
      return oldState.withMutations(state => {
        applyCost(state, action.cost);

        increaseValue(state, action.name);

        pushToHistory(state, action);
      });
    case constants.SKILLS_SPECIALITY_BUY:
      return oldState.withMutations(state => {
        applyCost(state, action.cost);

        addSpeciality(state, action.skill, action.speciality);

        pushToHistory(state, action);
      });
    default:
  }
  return oldState;
}
