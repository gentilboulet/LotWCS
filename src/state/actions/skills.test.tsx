import * as actions from './skills';

describe('Testing for skill action creators', () => {
  const noCost = {
    destiny: 0,
    discountIdx: -1, discountNewValue: 0,
    entanglement: 0
  };

  it('should create a buy skill action', () => {
    const skill = 'Awareness';
    const a = actions.skillsBuy(skill, noCost);
    expect( a.name ).toBe(skill);
  });

  it('should not create a buy invalid skill action', () => {
    const skill = 'Basketweaving';
    expect( () => actions.skillsBuy(skill, noCost) ).toThrow();
  });

  it('should create a buy speciality action', () => {
    const skill = 'Awareness';
    const speciality = 'Hear';
    const a = actions.skillSpecialityBuy(skill, speciality, noCost);
    expect( a.skill ).toBe(skill);
    expect( a.speciality ).toBe(speciality);
    expect( a ).toMatchSnapshot();
  });

  it('should not create a buy speciality action on an invalid skill', () => {
    const skill = 'Not a skill';
    const speciality = 'Hear';
    expect( () => actions.skillSpecialityBuy(skill, speciality, noCost) ).toThrow();
  });

  it('should not create a buy speciality action on an invalid speciality', () => {
    const skill = 'Awareness';
    const speciality = '';
    expect( () => actions.skillSpecialityBuy(skill, speciality, noCost) ).toThrow();
  });
});
