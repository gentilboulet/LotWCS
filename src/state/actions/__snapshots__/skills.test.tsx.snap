// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing for skill action creators should create a buy speciality action 1`] = `
Object {
  "cost": Object {
    "canPay": true,
    "destiny": 0,
    "original": 0,
  },
  "skill": "Awareness",
  "speciality": "Hear",
  "type": "SKILLS_SPECIALITY_BUY",
}
`;
